Android/app/src/main/java/app/intra/net/socks/OverrideSocksHandler.java

```diff
@@ -142,37 +143,41 @@ public void doConnect(Session session, CommandMessage commandMessage) throws IOE
 
     StatsListener listener = null;
     try {
-      listener = runPipes(upload, download, remoteServerPort, httpsTimeoutMs(tcpHandshakeMs));
-      // We consider a termination event as a potentially recoverable failure if
-      // (1) It is HTTPS
-      // (2) Some data has been uploaded (i.e. the TLS ClientHello)
-      // (3) No data has been received
-      // In this situation, the listener's uploadBuffer should be nonempty.
-      if (remoteServerPort == HTTPS_PORT && listener.uploadBytes > 0 && listener.downloadBytes == 0
-          && listener.uploadBuffer != null) {
-        // To attempt recovery, we
-        // (1) Close the remote socket (which has already failed)
-        // (2) Open a new one
-        // (3) Connect the new remote socket to the existing StreamPipe
-        // (4) Retry the ClientHello via splitRetry
-        socket.close();
-        download.stop();
-        socket = new Socket(remoteServerAddress, remoteServerPort);
-        socket.setTcpNoDelay(true);
-        socketDownload = socket.getInputStream();
-        socketUpload = socket.getOutputStream();
-
-        download = new StreamPipe(socketDownload, sessionDownload, DOWNLOAD);
-        download.setBufferSize(BUFFER_SIZE);
-
-        // upload is blocked in a read from sessionUpload, which cannot be canceled.
-        // However, we have modified StreamPipe to have a setter for the destination stream, so
-        // that the next transfer will be uploaded into the new socket instead of the old one.
-        upload.setDestination(socketUpload);
-
-        // Retry the connection while splitting the first flight into smaller packets.
-        // The retry doesn't trigger further retries if it fails, so it doesn't apply a timeout.
-        listener = splitRetry(upload, download, socketUpload, listener);
+      if (alwaysSplitHttps && remoteServerPort == HTTPS_PORT) {
+        listener = splitDirect(upload, download, sessionUpload, socketUpload);
+      } else {
+        listener = runPipes(upload, download, remoteServerPort, httpsTimeoutMs(tcpHandshakeMs));
```

Comment:
Maybe this section could be decomposed into a `splitOnFailure` method.

Reply:
I tried to do this, but it doesn't appear to be possible because the code further down needs to close the new `Socket` that was created in this section.  Java doesn't have pointers or multiple return, so there's no easy way to move this into a function while also replacing the caller's values for `upload`, `socket`, etc.